### PR TITLE
Add ClientAuthPolicy with ability to authenticate as a forwarded user

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR introduces a refactored and extended `ClientAuthPolicy` which now allows for the authentication "on behalf of" a user within an `auth_client`'s associated authority.

None of this is hooked up yet; next steps would be:

* Create an interstitial AuthPolicy that can delegate between `TokenAuthenticationPolicy` and `AuthClientPolicy` for API endpoints, always preferring Token auth when available
* Determine impact of API endpoints (e.g. blanket integration of this would allow auth_clients access to endpoints not previously accessible with auth_clients; this needs to be managed carefully)
* Update views to have different access rules/effective_principals

What this will ultimately accomplish:

* Individual API endpoints can support _either_ Token or `auth_client` authentication, not only one or the other
* For LMS, the LMS app will be able to create private groups _as_ users within their authority (using the already extant create-group endpoint and resulting in the correct creator_id)

This diff looks enormous (better part of 500 lines), but I'd wager <=75 lines of actual code; the rest is documentation and tests.

### CODE REVIEW

**Comment by seanh** [comment]
Geez, this stuff is complicated. Thanks for fighting through it all.

I want to summarize the `AuthClientPolicy` here, just as a way of verifying my understanding / echoing it back to you @lyzadanger:

* `AuthClientPolicy:`
  * `BasicAuthAuthenticationPolicy` (composition not inheritance)
  * `unauthenticated_userid():`
    * Returns `forwarded_userid()` (if there is one) or `BasicAuthAuthenticationPolicy.unauthenticated_userid()` (which may be `None`)
    * This is an _un_authenticated userid: If there's a userid in an `X-Forwarded-User` header or in the `username` part of a Basic HTTP auth header then it's blindly returned, regardless of whether that user exists in the DB or is within the authclient's authority or whether the authclient id exists in the DB or the `client_id` matched the `client_secret`
  * `authenticated_userid():`
    * Returns `forwarded_userid()` is there is one and it's valid, or `None`
    * Unlike `unauthenticated_userid()`, `authenticated_userid()` never returns `client_id`s
    * Also unlike `unauthenticated_userid()`, `authenticated_userid()` actually verifies that the `userid` exists in the `User` table in the DB, that the user's authority matches the authclient's, and that the `client_secret` for the `authclient` is correct.
    * None of this verification is done in `AuthClientPolicy` itself though, it's done by the `check()` function (called indirectly as `BasicAuthAuthenticationPolicy.callback()`) that's passed in to `AuthClientPolicy.__init__()`.
    * Since `AuthClientPolicy` isn't hooked up yet nothing outside of the tests actually constructs an `AuthClientPolicy` yet so this passing in of the `check` function isn't done yet.
    * But the function that's intended to be passed in as the `check` is `check_auth_client()`, which is built on top of two separate functions `verify_auth_client()` and `verify_forwarded_user()`.
 
It makes sense that `unauthenticated_userid` can return `client_id`s but `authenticated_userid` can't, because when we're doing authenticated-as-the-authclient-itself we don't want `request.authenticated_userid` or `request.user` to be set.

It _looks_ as if, when doing authenticated-as-the-authclient-itself, that nothing checks that the `client_id` actually exists in the DB and that the `client_secret` is correct. But in fact [`BasicAuthAuthenticationPolicy` subclasses `CallbackAuthenticationPolicy`](https://github.com/Pylons/pyramid/blob/5b1e29544ff881bed4fbdfc48f32c67c19554035/pyramid/authentication.py#L1048) and [`CallbackAuthenticationPolicy`'s `effective_principals()` method calls the `callaback()` function](https://github.com/Pylons/pyramid/blob/5b1e29544ff881bed4fbdfc48f32c67c19554035/pyramid/authentication.py#L159-L163) (in our case, `check_auth_client()`) :heavy_check_mark: 

I have a couple of high-level suggestions that I want to get your response to first:

 ## Move `check_auth_client()` into `AuthClientPolicy`?

I wonder if `check_auth_client()` belongs in the `AuthClientPolicy` class itself, rather than as utility functions? Just because they seem to be intended only to be used by `AuthClientPolicy`, and it might be easier to follow the code if it's all kept together in one class. Something like:

```python
@interface.implementer(interfaces.IAuthenticationPolicy)
class AuthClientPolicy(object):
    def __init__(self):
        self.basic_auth_policy = BasicAuthAuthenticationPolicy(check=self._check)
        
    def _check(self, username, password, request):
        # (Calls self._verify_auth_client() and self._verify_forwarded_user())
        
    def _verify_auth_client(...):
        ...
    
    def _verify_forwarded_user(...):
        ...

    ...

# You don't need to pass any `check` in when constructing it,
# as that's handled internally now.
auth_client_policy = AuthClientPolicy()
```

(`principals_for_auth_client()` might also belong in `AuthClientPolicy` one day for the same reason.)

I guess the downside of this is that now you can't unit test `verify_auth_client()`, `verify_forwarded_user()`, `check_auth_client()`, and `AuthClientPolicy` separately, as they're all one thing. So then I wonder if the functions could remain in `util` but be imported and called by `AuthClientPolicy` directly rather than using a `check` arg, and then `AuthClientPolicy` tests could patch them:

```python
from h.auth.util import check_auth_client

@interface.implementer(interfaces.IAuthenticationPolicy)
class AuthClientPolicy(object):
    def __init__(self):
        self.basic_auth_policy = BasicAuthAuthenticationPolicy(check=check_auth_client)

    ...

# You don't need to pass any `check` in when constructing it,
# as that's handled internally now.
auth_client_policy = AuthClientPolicy()
```

The only thing I don't like about this is that `check_auth_client`, being only intended to be used by `AuthClientPolicy`, isn't really a "util" or "helper", because I'd normally think of utils or helpers as being called by lots of places, but not sure where else you'd put it.

## Split AuthClientPolicy into AuthClientPolicy and XForwardedUserPolicy?

I wonder if authenticating as an authclient and authenticating as a user via and authclient using `X-Forwarded-User` are two different things? I wonder how it would look if these were two different auth policies?

**Comment by lyzadanger** [comment]
I like your idea about moving `check_auth_client` into `AuthClientPolicy` itself and auto-assigning it as the `check` callback. It could make all of this easier to parse. I _think_ I should be able to move tests over OK.

As an aside, both the `policy` and `util` modules are getting unwieldy—I'm quite aware of that (and it makes testing a pain because the test modules are so complex). In a subsequent PR, I'd like to look at moving policies into individual modules; plus there's crap in util that will go away once view-level logic is removed for auth_client endpoints.

What I'll do now:

* Move `check_auth_client` function to be a method of `AuthClientPolicy`
* Make it possible to instantiate `AuthClientPolicy` without passing a check callback
* Improve commenting in some areas

Let's see if that feels more coherent and easier to follow?

**Comment by seanh** [comment]
Ok, let me know when you want me to take another look

**Comment by lyzadanger** [comment]
@seanh This is ready for re-review.

* I've re-factored the check callback as a class method on `AuthClientPolicy` and defaulted to use that as the `check` callback for `AuthClientPolicy` instances.
* I've expanded commenting in a few places.
* I've also changed `forwarded_userid` to be a class method as it seems more coherent here

File: h/auth/policy.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from pyramid.authentication import CallbackAuthenticationPolicy
 from zope import interface
 
+from h.auth import util
```
Comment (lyzadanger) [inline h/auth/policy.py]: Moving `check` into the `AuthClientPolicy` class means I have multiple dependencies on the util functions it invokes. I hope this is OK?

File: h/auth/policy.py [hunk 2]
```diff
@@ -41,34 +43,104 @@ def forget(self, request):
...
+
+        If the ``check`` callback returns None—that is, if authentication is
+        unsuccessful—the returned principals will only contain Pyramid's
+        ``system.Everyone`` principal (and Pyramid will not consider the
+        request as authenticated).
+
+        :rtype: list ``['system.Everyone']`` concatenated with any principals
+                from a successful authentication, if any
+        """
+        return self.basic_auth_policy.effective_principals(request)
```
Comment (lyzadanger) [inline h/auth/policy.py]: I took a look at re-implementing `effective_principals` instead of proxying to `BasicAuthAuthenticationPolicy`'s variant (inherited through `CallbackAuthenticationPolicy`), but there's just enough Pyramid magic going on there that it seems unwise.

File: h/auth/policy.py [hunk 3]
```diff
@@ -41,34 +43,104 @@ def forget(self, request):
...
+    a request with an auth_client forwarded user, the latter has an additional
+    principal, ``client:{client_id}@{authority}`` to mark it as being authenticated
+    on behalf of an auth_client
     """
 
-    def __init__(self, check, realm='Realm', debug=False):
-        self.check = check
-        self.realm = realm
-        self.debug = debug
+    def __init__(self, check=None):
```
Comment (lyzadanger) [inline h/auth/policy.py]: `AuthClienPolicy()` will still accept a `check` param. TBH, this was mostly to make testing more sane, so I hope it's not a cop-out.

File: h/auth/policy.py [hunk 4]
```diff
@@ -41,34 +43,104 @@ def forget(self, request):
...
+        if check is None:
+            check = AuthClientPolicy.check
```
Comment (lyzadanger) [inline h/auth/policy.py]: TIL: you can't default a constructor param for to a class method on the class for which you're creating an instance :smile:

File: tests/h/auth/policy_test.py [hunk 5]
```diff
@@ -113,43 +113,232 @@ def test_forget_uses_api_policy_for_api_paths(self, api_request):
...
+    def test_unauthenticated_userid_doesnt_proxy_to_basic_auth_if_forwarded_user(self,
+                                                                                auth_policy,
+                                                                                pyramid_request,
+                                                                                BasicAuthAuthenticationPolicy):
+        pyramid_request.headers['X-Forwarded-User'] = 'dingbat'
+        auth_policy.basic_auth_policy = BasicAuthAuthenticationPolicy(check=AuthClientPolicy.check)
+
+        auth_policy.unauthenticated_userid(pyramid_request)
+
+        auth_policy.basic_auth_policy.unauthenticated_userid.assert_not_called()
```
Comment (seanh) [inline tests/h/auth/policy_test.py]: This land mine again, `assert_not_called()` doesn't exist in the Python 2 / third-party mock library! And `auth_policy.basic_auth_policy` is an unconstrained mock. The `BasicAuthAuthenticationPolicy` mock is created using autospeccing (via our `patch` fixture, which uses autospec) but unfortunately [autospec doesn't spec the return values of mock objects and their methods](https://hyp.is/cfYO-JdGEeiL-P8E8GQW1g/www.seanh.cc/posts/the-problem-with-mocks) (it can't, because mock would have to actually call our methods in order to see what shape of object they return). So while the `BasicAuthAuthenticationPolicy` is autospec'd, object instances of that class (i.e. the mock class's return values) are unconstrained mocks.

You might think autospec could maybe detect the thing being mocked is a class and autospec the class's return value / the object instances as well, but unfortunately there are reasons why that would cause problems.

The tests can fix this like this by doing mock's work for it:

```python
@pytest.fixture
def BasicAuthAuthenticationPolicy(self, patch):
    BasicAuthAuthenticationPolicy = patch("h.auth.policy.BasicAuthAuthenticationPolicy")
    BasicAuthAuthenticationPolicy.return_value = mock.create_autospec(BasicAuthAuthenticationPolicy, instance=True, spec_set=True)
```

File: h/auth/policy.py (hunk 3 above)
Comment (lyzadanger) [reply h/auth/policy.py]: I'm sorry—I'm going absolutely bananas trying to get this to work in tests. It's becoming a mess. I'm going to have to set it aside for the moment to try to get to the more key items here...

File: h/auth/policy.py [hunk 6]
```diff
@@ -41,34 +41,89 @@ def forget(self, request):
...
+    principal, ``client:{client_id}@{authority}`` to mark it as being authenticated
+    on behalf of an auth_client
     """
 
-    def __init__(self, check, realm='Realm', debug=False):
+    def __init__(self, check):
+        """
+        :param: check A callback function that accepts (username, password, request)
+                and should return a list of principals or `None` if auth
+                unsucessful; it will be called with the username and password
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply h/auth/policy.py]: Sorry about the rSt misses in this PR. I am continuing to find it more complex than I'd like, but I'll buckle down and review all my docstrings.

File: h/auth/policy.py [hunk 7]
```diff
@@ -41,34 +41,89 @@ def forget(self, request):
...
-        However, the request will be authenticated. In a view,
-        ``effective_principals=security.Authenticated`` would be satisfied.
+        :rtype: str
+        """
+        proxy_userid = self.forwarded_userid(request)
+        if proxy_userid is not None:
+            return proxy_userid
+
+        # username from BasicAuth header
+        return self.basic_auth_policy.unauthenticated_userid(request)
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply h/auth/policy.py]: I totally agree on the `proxy` naming; what was I thinking? :)

File: h/auth/policy.py [hunk 8]
```diff
@@ -41,34 +41,89 @@ def forget(self, request):
...
-    def __init__(self, check, realm='Realm', debug=False):
+    def __init__(self, check):
+        """
+        :param: check A callback function that accepts (username, password, request)
+                and should return a list of principals or `None` if auth
+                unsucessful; it will be called with the username and password
+                parsed from the Authentication header
+        """
+        self.basic_auth_policy = BasicAuthAuthenticationPolicy(check=check)
         self.check = check
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply h/auth/policy.py]: Already fixed in my updated PR because I noticed it, too.

File: h/auth/policy.py [hunk 9]
```diff
@@ -78,6 +150,57 @@ def forget(self, request):
...
+        an auth_client record in the DB.
+
+        If an HTTP `X-Forwarded-User` header is present in the request, this
+        represents the intent to authenticate "on behalf of" a user within
+        the auth_client's authority. If this header is present, the user indicated
+        by its value (a `~h.models.user.User.userid`) _must_ exist and be within
+        the auth_client's authority, or authentication will fail.
+
+        This function is intended to be used as a callback for an
+        authentication policy
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply h/auth/policy.py]: Good catch.

File: h/auth/policy.py [hunk 10]
```diff
@@ -78,6 +150,57 @@ def forget(self, request):
         """Not implemented for basic auth client policy."""
         return []
 
+    @classmethod
+    def check(cls, username, password, request):
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply h/auth/policy.py]: I will change them to static.

I concur that `_forwarded_user_id` is unambiguously a helper function, but I think there is a (small, admittedly) argument that `check` could be public; it feels quite standalone in my tests.

That is, you can do `AuthClientPolicy.check(request)` on any ol' Pyramid request without instantiating an instance of the policy and it will return the correct principals (or `None`).

**Comment by lyzadanger** [comment]
Whoof! This is getting wild. @seanh I think it's ready for a re-review. I borked history and managed to lose the fixup commit for the changes since your review; they've been squashed into the existing commits (first time I've messed up a rebase in a long time; sorry about that; long week).

Some highlights:

* `AuthClientPolicy._forwarded_userid` is now a private `@staticmethod`
* `AuthClientPolicy.check` is now a `@staticmethod`
* `AuthClientPolicy._basic_auth_policy` is now a private attribute, and tests have been widely updated to not test things on this attribute directly
* `h.auth.util.verify_forwarded_user` has been removed, and the user fetching and checking is now part of `AuthClientPolicy.check`
* I've gone over every docstring and made it better and tested it

File: h/auth/util.py [hunk 11]
```diff
@@ -157,25 +149,42 @@ def check_auth_client(username, password, request):
...
+    :type client: :py:class:`h.models.auth_client.AuthClient`
+    :rtype: list
+    """
+    user_principals = principals_for_user(user)
+    client_principals = principals_for_auth_client(client)
+
+    all_principals = user_principals + client_principals
+    distinct_principals = list(set(all_principals))
+
+    return distinct_principals
```
Comment (seanh) [inline h/auth/util.py]: I'm wondering why an `X-Forwarded-User` needs the principals for both the user and the auth client, instead of just the principals for the user?

File: h/auth/util.py [hunk 12]
```diff
@@ -157,25 +149,42 @@ def check_auth_client(username, password, request):
...
 
-    :type client: h.models.AuthClient
+    :type client: :py:class:`h.models.auth_client.AuthClient`
     :rtype: list
     """
 
     principals = set([])
 
-    principals.add('auth_client:{authority}'.format(authority=client.authority))
+    principals.add('client:{client_id}@{authority}'.format(client_id=client.id, authority=client.authority))
```
Comment (seanh) [inline h/auth/util.py]: Why the change from `auth_client:` to `client:`?

File: h/auth/util.py (hunk 11 above)
Comment (lyzadanger) [reply h/auth/util.py]: I want to be able to "mark" a request as having been authenticated via an auth_client. Otherwise there is no easy way to discern a forwarded-user auth'd request from a Token auth'd request.

File: h/auth/util.py (hunk 12 above)
Comment (lyzadanger) [reply h/auth/util.py]: `auth_client` felt H-jargon-y to me; I can change it back if you like!
